File: src/DotNetEveryDay.Extensions/Extensions/AttributeExtensions.fs
namespace DotNetEveryDay.Extensions.Extensions

open System.Collections.Concurrent
open System.Runtime.CompilerServices
open System

[<AllowNullLiteral>]
[<Extension>]
type AttributeExtensions() =
    [<DefaultValue>]
    static val mutable private CachedAttributes: ConcurrentDictionary<string, Attribute>
    
    static do
        AttributeExtensions.CachedAttributes <- ConcurrentDictionary<string, Attribute>()
    
    [<Extension>]
    static member GetAttribute<'TAttribute when 'TAttribute :> Attribute>(_type: Type, propertyName: string) =     
        AttributeExtensions.CachedAttributes.GetOrAdd(
            $"{_type}.{propertyName}_{typeof<'TAttribute>}",
            fun i -> AttributeExtensions.GetAttributeFromProperty<'TAttribute>(_type, propertyName))
        :?> 'TAttribute
        
    static member private GetAttributeFromProperty<'TAttribute when 'TAttribute :> Attribute>(
        _type: Type, propertyName: string) : Attribute =
        let attribute  =
            _type.GetProperty(propertyName).GetCustomAttributes(typeof<'TAttribute>, false)
            |> Seq.tryHead

        match attribute with
        | Some value -> value :?> Attribute
        | None -> null

File: src/DotNetEveryDay.Extensions/Extensions/AttributeExtensions.cs
using System.Collections.Concurrent;

namespace DotNetEveryDay.Extensions.Extensions;


public static class AttributeExtensions
{
    private static readonly ConcurrentDictionary<string, Attribute?> CachedAttributes =
        new(Environment.ProcessorCount, 256);

    /// <summary>
    /// Get attribute if exist or null for property by property name
    /// </summary>
    /// <param name="type"></param>
    /// <param name="propertyName"></param>
    /// <typeparam name="TAttribute"></typeparam>
    /// <returns></returns>
    public static TAttribute? GetPropertyAttribute<TAttribute>(this Type type, string propertyName)
        where TAttribute : Attribute
    {
        if (type == null) throw new ArgumentNullException(nameof(type));

        if (propertyName.IsNullOrWhiteSpace()) return null;
        
        return (TAttribute?) CachedAttributes.GetOrAdd($"{type}.{propertyName}_{typeof(TAttribute)}",
            _ => (TAttribute?) type.GetProperty(propertyName)
            ?.GetCustomAttributes(typeof(TAttribute), false)
            .FirstOrDefault());
    }
}